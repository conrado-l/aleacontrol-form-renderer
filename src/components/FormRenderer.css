.form-renderer {
    padding: 10px 15px;
    background-color: #ffffff;
    box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
}
