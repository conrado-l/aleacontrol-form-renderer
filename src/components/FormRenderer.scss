@import "variables";

.form-renderer {
    padding: 10px 15px;
    background-color: #ffffff;
    box-shadow: $form-box-shadow-horizontal, $form-box-shadow-vertical;
}
