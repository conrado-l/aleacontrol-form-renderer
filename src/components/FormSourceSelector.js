import React, { Component } from 'react'
import Button from 'react-bootstrap/Button'
import '../styles/FormSourceSelector.scss'
import PropTypes from 'prop-types'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'

/**
 * Button that should change the current form
 */
class FormSourceSelector extends Component {
  render () {
    return (
      <Row>
        <Col xs={12} md={8} xl={6}>
          <div className='form-source-selector w-100 mt-2 d-flex align-content-center justify-content-around'>
            <Button variant='outline-info' className='change-data-button' onClick={this.props.changeFormSample}>
                Change Form Sample</Button>
          </div>
        </Col>
      </Row>
    )
  }
}

FormSourceSelector.propTypes = {
  changeFormSample: PropTypes.func
}
export default FormSourceSelector
