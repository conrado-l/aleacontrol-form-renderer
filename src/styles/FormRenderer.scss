@import "variables";
@import "mixins";

.form-renderer {
    padding: 10px 15px;
    background-color: $box-background-color;
    border-radius: $box-border-radius;
    @include box-shadow;
}
