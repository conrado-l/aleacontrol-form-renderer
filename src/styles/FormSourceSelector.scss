@import "variables";
@import "mixins";

.form-source-selector {
  height: 60px;
  width: 300px;
  padding: 5px;
  background-color: $box-background-color;
  border-radius: $box-border-radius;
  @include box-shadow;
}

.change-data-button {
  height: 50px;
}
