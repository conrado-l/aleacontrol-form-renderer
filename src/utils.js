/**
 * Transforms a JSON to an object
 * @param json
 * @returns {ParsedPath | ParsedUrlQuery | * | UrlWithStringQuery | UrlWithParsedQuery | Url | number}
 */
export function JSONToObject (json) {
  return JSON.parse(JSON.stringify(json))
}
