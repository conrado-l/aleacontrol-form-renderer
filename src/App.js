import React, { Component } from 'react'
import FormRenderer from './components/FormRenderer'
import './App.css'

class App extends Component {
  render () {
    return (
      <div className='App'>
        <FormRenderer />
      </div>
    )
  }
}

export default App
