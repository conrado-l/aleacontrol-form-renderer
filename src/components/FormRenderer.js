import React, {Component} from 'react'
import FormGroup from './FormGroup'
import SelectInput from './SelectInput'
import StringInput from './StringInput'
import EmailInput from './EmailInput'
import RadioInput from './RadioInput'
import PhoneInput from './PhoneInput'
import MultiStringInput from './MultiStringInput'
import Button from 'react-bootstrap/Button';
import './FormRenderer.scss'
import Form from 'react-bootstrap/Form'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'

class FormRenderer extends Component {
    constructor(props) {
        super(props);

        this.state = {
            form: [
                {
                    'label': 'First name',
                    'name': 'first_name',
                    'value': null,
                    'component': 'string',
                    'col': 6
                },
                {
                    'label': 'Last name',
                    'name': 'last_name',
                    'value': null,
                    'component': 'string',
                    'col': 6
                },
                {
                    'label': 'Email',
                    'name': 'email',
                    'value': null,
                    'component': 'email',
                    'col': 6
                },
                {
                    'label': 'Gender',
                    'name': 'gender',
                    'value': 'male',
                    'options': [
                        {
                            description: 'Male',
                            value: 'male'
                        },
                        {
                            description: 'Female',
                            value: 'female'
                        }
                    ],
                    'component': 'radio',
                    'col': 6
                },
                {
                    'label': 'Continent',
                    'name': 'continent',
                    'value': null,
                    'options': [
                        {
                            description: 'Europe',
                            value: 'europe'
                        },
                        {
                            description: 'America',
                            value: 'america'
                        },
                        {
                            description: 'Asia',
                            value: 'asia'
                        },
                        {
                            description: 'Africa',
                            value: 'africa'
                        },
                        {
                            description: 'Australia',
                            value: 'australia'
                        },
                        {
                            description: 'Antartica',
                            value: 'antartica'
                        }
                    ],
                    'component': 'select',
                    'col': 12
                },
                {
                    'label': 'Phone number',
                    'name': 'phone_number',
                    'value': {
                        ext: '+387',
                        phone: '123456'
                    },
                    'component': 'phone',
                    options: [
                        {
                            value: '387',
                            description: 'Bosnia and Herzegovina'
                        },
                        {
                            value: '54',
                            description: 'Argentina'
                        }
                    ],
                    col: 6
                },
                {
                    'label': 'Interests',
                    'name': 'interests',
                    'value': ['Coding', 'Playing piano', 'Product management'],
                    'component': 'multistring',
                    'col': 6
                }
            ],
            request: {}
        };

        this.sendForm = this.sendForm.bind(this)
        this.updateFormInput = this.updateFormInput.bind(this)
        this.changeSampleData = this.changeSampleData.bind(this)
    }

    sample2 = [
        {
            'label': 'Brand',
            'name': 'brand',
            'value': 'Nike',
            'component': 'string',
            'col': 6
        },
        {
            'label': 'Model',
            'name': 'model',
            'value': 'Total 90',
            'component': 'string',
            'col': 6
        },
        {
            'label': 'Email',
            'name': 'email',
            'value': 'conrado.lk@gmail.com',
            'component': 'email',
            'col': 6
        }
    ]

    // Mapping: JSON component name attribute -> component imports
    components = {
        'string': StringInput,
        'email': EmailInput,
        'radio': RadioInput,
        'select': SelectInput,
        'phone': PhoneInput,
        'multistring': MultiStringInput
    }

    updateFormInput(inputKey, inputValue) {
        const newRequestState = {...this.state.request};
        newRequestState[inputKey] = inputValue;
        this.setState({
                request: newRequestState
            }
        )
    }

    sendForm() {
        console.log(this.state.request)
    }

    changeSampleData() {
        this.setState({
                form: sample2
            }
        )
    }

    /**
     * Renders the inputs dynamically according to the mapping
     * @returns {any[]}
     */
    renderInputs() {
        return this.state.form.map((input) => {

            const DynamicInputTag = this.components[input.component]

            return <FormGroup key={input.name} label={input.label}>
                <DynamicInputTag {...input} update={this.updateFormInput}/>
            </FormGroup>
        })
    }

    render() {

        return (
            <Form className='form-renderer rounded p-3 my-sm-5 col-md-6 col-sm-6 col-xs-12 col-xl-3'>
                {/*<Row>*/}
                    {/*<Col md={6} sm={6} xs={6} xl={3}>*/}
                        {this.renderInputs()}
                        <hr/>
                        <div className='d-flex justify-content-around'>
                            <Button variant='outline-info' onClick={this.changeSampleData}> Change Data</Button>
                            <Button variant='outline-primary' onClick={this.sendForm}> Send Form</Button>
                        </div>
                    {/*</Col>*/}
                {/*</Row>*/}
            </Form>
        )
    }
}

export default FormRenderer
