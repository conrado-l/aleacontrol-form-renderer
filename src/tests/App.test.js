import React from 'react'
import { mount } from 'enzyme'
import App from '../App'
import renderer from 'react-test-renderer'
import FormRenderer from '../components/FormRenderer'
import FormSourceSelector from '../components/FormSourceSelector'
import SampleForm1 from '../sample1'
import SampleForm2 from '../sample2'
import { JSONToObject } from '../utils'

describe('App', () => {
  it('should render correctly app component', () => {
    const AppComponent = renderer
      .create(<App />)
      .toJSON()
    expect(AppComponent).toMatchSnapshot()
  })

  it('should render the form renderer component', () => {
    const AppComponent = mount(<App />)
    expect(AppComponent.find(FormRenderer)).toHaveLength(1)
  })

  it('should render the form source selector component', () => {
    const AppComponent = mount(<App />)
    expect(AppComponent.find(FormSourceSelector)).toHaveLength(1)
  })

  it('should have the parsed file sample1.json as initial/mounted state', () => {
    const AppComponent = mount(<App />)
    const parsedSampleForm1 = JSONToObject(SampleForm1)

    expect(AppComponent.state().form).toEqual(parsedSampleForm1)
  })

  it('should have the parsed file sample2.json as state when the changeSampleData executes', () => {
    const AppComponent = mount(<App />)
    const AppComponentInstance = AppComponent.instance()
    const parsedSampleForm2 = JSONToObject(SampleForm2)

    AppComponentInstance.changeFormSample()

    expect(AppComponent.state().form).toEqual(parsedSampleForm2)
  })
})
