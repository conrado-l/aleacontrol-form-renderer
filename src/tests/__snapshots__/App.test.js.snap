// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly app component 1`] = `
<div
  className="container"
>
  <div
    className="center-content"
  >
    <div
      className="row"
    >
      <div
        className="col-xl-6 col-md-8 col-12"
      >
        <div
          className="form-source-selector w-100 mt-2 d-flex align-content-center justify-content-around"
        >
          <button
            className="change-data-button btn btn-outline-info"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Change Form Sample
          </button>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-xl-6 col-md-8 col-12"
      >
        <form
          className="form-renderer p-3 my-3"
        >
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              First name
            </label>
            <div>
              <div
                className="row"
              >
                <div
                  className="col-sm-6 col-12"
                >
                  <input
                    className="form-control"
                    name="first_name"
                    onChange={[Function]}
                    type="text"
                    value="Paul"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              Last name
            </label>
            <div>
              <div
                className="row"
              >
                <div
                  className="col-sm-6 col-12"
                >
                  <input
                    className="form-control"
                    name="last_name"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              Email
            </label>
            <div>
              <div
                className="row"
              >
                <div
                  className="col-sm-8 col-12"
                >
                  <input
                    className="form-control"
                    name="email"
                    onChange={[Function]}
                    type="email"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              Gender
            </label>
            <div>
              <div
                className="radio-input-container"
              >
                <div>
                  <div
                    className="form-check form-check-inline"
                  >
                    <input
                      className="form-check-input"
                      defaultChecked={true}
                      disabled={false}
                      id="male"
                      name="gender"
                      onChange={[Function]}
                      type="radio"
                      value="male"
                    />
                    <label
                      className="form-check-label"
                      htmlFor="male"
                      title=""
                      type="checkbox"
                    >
                      Male
                    </label>
                  </div>
                </div>
                <div>
                  <div
                    className="form-check form-check-inline"
                  >
                    <input
                      className="form-check-input"
                      defaultChecked={false}
                      disabled={false}
                      id="female"
                      name="gender"
                      onChange={[Function]}
                      type="radio"
                      value="female"
                    />
                    <label
                      className="form-check-label"
                      htmlFor="female"
                      title=""
                      type="checkbox"
                    >
                      Female
                    </label>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              Continent
            </label>
            <div>
              <div
                className="row"
              >
                <div
                  className="col-sm-12 col-12"
                >
                  <select
                    className="form-control"
                    defaultValue="america"
                    onChange={[Function]}
                  >
                    <option
                      value="europe"
                    >
                       
                      Europe
                    </option>
                    <option
                      value="america"
                    >
                       
                      America
                    </option>
                    <option
                      value="asia"
                    >
                       
                      Asia
                    </option>
                    <option
                      value="africa"
                    >
                       
                      Africa
                    </option>
                    <option
                      value="australia"
                    >
                       
                      Australia
                    </option>
                    <option
                      value="antartica"
                    >
                       
                      Antartica
                    </option>
                  </select>
                </div>
              </div>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              Phone number
            </label>
            <div>
              <div
                className="row"
              >
                <div
                  className="col-sm-6 col-12"
                >
                  <div
                    className="row"
                  >
                    <div
                      className="col-12"
                    >
                      <select
                        className="form-control"
                        defaultValue="+387"
                        onChange={[Function]}
                      >
                        <option
                          value="387"
                        >
                           
                          Bosnia and Herzegovina
                        </option>
                        <option
                          value="54"
                        >
                           
                          Argentina
                        </option>
                      </select>
                    </div>
                  </div>
                </div>
                <div
                  className="col-sm-6 col-12"
                >
                  <div
                    className="row"
                  >
                    <div
                      className="col-12"
                    >
                      <input
                        className="form-control"
                        name="phone"
                        onChange={[Function]}
                        type="text"
                        value="123456"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              Interests
            </label>
            <div>
              <div>
                <div
                  className="mb-3"
                >
                  <div
                    className="row"
                  >
                    <div
                      className="col-sm-6 col-12"
                    >
                      <input
                        className="form-control"
                        name="0"
                        onChange={[Function]}
                        type="text"
                        value="Coding"
                      />
                    </div>
                  </div>
                </div>
                <div
                  className="mb-3"
                >
                  <div
                    className="row"
                  >
                    <div
                      className="col-sm-6 col-12"
                    >
                      <input
                        className="form-control"
                        name="1"
                        onChange={[Function]}
                        type="text"
                        value="Playing piano"
                      />
                    </div>
                  </div>
                </div>
                <div
                  className="mb-3"
                >
                  <div
                    className="row"
                  >
                    <div
                      className="col-sm-6 col-12"
                    >
                      <input
                        className="form-control"
                        name="2"
                        onChange={[Function]}
                        type="text"
                        value="Product management"
                      />
                    </div>
                  </div>
                </div>
                <div
                  className="d-flex justify-content-around"
                >
                  <button
                    className="btn btn-outline-success"
                    disabled={false}
                    id="add"
                    onClick={[Function]}
                    type="button"
                  >
                     Add input
                  </button>
                  <button
                    className="btn btn-outline-danger"
                    disabled={false}
                    id="delete"
                    onClick={[Function]}
                    type="button"
                  >
                     Delete last input 
                  </button>
                </div>
              </div>
            </div>
          </div>
          <hr />
          <div
            className="d-flex justify-content-around"
          >
            <button
              className="btn btn-outline-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
               Send Form 
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
